packages/bitcoin-dlc-provider/lib/BitcoinDlcProvider.ts

```diff
@@ -1806,81 +1705,387 @@ Payout Group not found',
+            .serialize()
+            .slice(1),
+          value: Number(input.prevTx.outputs[input.prevTxVout].value.sats),
+        },
+        serialId: input.inputSerialId,
+      });
+    });
+
+    // sort all inputs in ascending order by serial ID
+    // The only reason we are doing this is for privacy. If the fundingInput is
+    // always first, it is very obvious. Hence, a serialId is randomly generated
+    // and the inputs are sorted by that instead.
+    const sortedPsbtInputs = psbtInputs.sort((a, b) =>
+      Number(a.serialId - b.serialId),
+    );
+
+    // Get index of fundingInput
+    const fundingInputIndex = sortedPsbtInputs.findIndex(
+      (input) => input.serialId === dlcClose.fundInputSerialId,
+    );
+
+    psbt.addOutput({
+      value: Number(dlcClose.offerPayoutSatoshis),
+      address: address.fromOutputScript(dlcOffer.payoutSPK, network),
+    });
+
+    psbt.addOutput({
+      value: Number(dlcClose.acceptPayoutSatoshis),
+      address: address.fromOutputScript(dlcAccept.payoutSPK, network),
+    });
+
+    // add to psbt
+    sortedPsbtInputs.forEach((input, i) => psbt.addInput(input));
+
+    const offerer = await this.isOfferer(dlcOffer, dlcAccept);
+
+    // Generate keypair to sign inputs
+    const fundPrivateKeyPair = await this.GetFundKeyPair(
+      dlcOffer,
+      dlcAccept,
+      offerer,
+    );
+
+    // Sign dlc fundinginput
+    psbt.signInput(fundingInputIndex, fundPrivateKeyPair);
+
+    const partialSig = [
+      {
+        pubkey: offerer ? dlcAccept.fundingPubKey : dlcOffer.fundingPubKey,
+        signature: dlcClose.closeSignature,
+      },
+    ];
+    psbt.updateInput(fundingInputIndex, { partialSig });
+
+    for (let i = 0; i < psbt.data.inputs.length; ++i) {
+      if (i === fundingInputIndex) continue;
+      if (!psbt.data.inputs[i].partialSig) psbt.data.inputs[i].partialSig = [];
+
+      const witnessI = dlcClose.fundingSignatures.witnessElements.findIndex(
+        (el) =>
+          Buffer.compare(
+            Script.p2wpkhLock(hash160(el[1].witness)).serialize().slice(1),
+            psbt.data.inputs[i].witnessUtxo.script,
+          ) === 0,
       );
 
-      psbt.signInput(0, fundPrivateKeyPair);
-      psbt.validateSignaturesOfInput(0);
+      const partialSig = [
+        {
+          pubkey:
+            dlcClose.fundingSignatures.witnessElements[witnessI][1].witness,
+          signature:
+            dlcClose.fundingSignatures.witnessElements[witnessI][0].witness,
+        },
+      ];
 
-      for (let i = 1; i < inputs.length + 1; i++) {
-        const wallet: Address = await this.getMethod('getWalletAddress')(
-          inputs[i - 1].address,
-        );
-        const keyPair = await this.getMethod('keyPair')(wallet.derivationPath);
-        psbt.signInput(i, keyPair);
-        psbt.validateSignaturesOfInput(i);
-      }
+      psbt.updateInput(i, { partialSig });
     }
 
-    return psbt;
+    psbt.validateSignaturesOfAllInputs();
```

Comment:
Might be good to have specific error for which input signature caused the psbt validation to fail

Reply:
This would automatically throw with that right?

Reply:
Yea this is something we can probably add later. Would just be nice to be able to see specific errors if close signature or funding inputs signatures were incorrect.